pkg/resourcemanager/controller/health/progressing/reconciler.go

```diff
@@ -173,19 +178,39 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, mr *resourc
 
 // checkProgressing checks whether the given object is progressing. It returns a bool indicating whether the object is
 // progressing, a reason for it if so and an error if the check failed.
-func checkProgressing(obj client.Object) (bool, string) {
+func (r *Reconciler) checkProgressing(ctx context.Context, obj client.Object) (bool, string, error) {
 	if obj.GetAnnotations()[resourcesv1alpha1.SkipHealthCheck] == "true" {
-		return false, ""
+		return false, "", nil
 	}
 
+	var (
+		progressing bool
+		reason      string
+	)
+
 	switch o := obj.(type) {
 	case *appsv1.Deployment:
-		return health.IsDeploymentProgressing(o)
+		progressing, reason = health.IsDeploymentProgressing(o)
+		if progressing {
+			return true, reason, nil
+		}
+
+		// health.IsDeploymentProgressing might return false even if there are still (terminating) pods in the system
+		// belonging to an older ReplicaSet of the Deployment, hence, we have to check for this explicitly.
+		exactNumberOfPods, err := health.DeploymentHasExactNumberOfPods(ctx, r.TargetClient, o)
+		if err != nil {
+			return progressing, reason, err
+		}
+		if !exactNumberOfPods {
+			return true, "there are still non-terminated old pods", nil
+		}
```

Comment:
No, we still call `health.IsDeploymentProgressing` first, and this will already return `progressing=true` in such situation (since this is well observable via the `Deployment`'s status). Only when the `Deployment`'s status tells that `progressing=false` we reach this point where we check that the deployment has exact number of pods - which can only mean that there are still non-terminated old pods. :)